src/encoding.ts

```diff
@@ -0,0 +1,349 @@
+import { HIGH_SURROGATE_MIN, HIGH_SURROGATE_MAX, LOW_SURROGATE_MIN, LOW_SURROGATE_MAX } from './string';
+
+type ByteBuffer = Uint8Array | Buffer | number[];
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function validateArguments(data: any) {
+	return data == null ? false : true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [ 0, Infinity ], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface Codec {
+	encode(data: string): number[];
+	decode(data: ByteBuffer): string;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export const ascii: Codec = {
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string): number[] {
+		if (!validateArguments(data)) {
```

Comment:
Since `validateArguments` is just testing for data to be null or undefined, just say

```js
if (data == null) {
   ..
}
```

and get rid of `validateArguments`.